### PR TITLE
Propose fix perceptual loss sqrt nan

### PR DESCRIPTION
**Pull request by cvbourne**
Fixes # 8412

### Description

This PR fixes a numerical stability issue in the PerceptualLoss implementation where the `normalize_tensor` function can produce NaN gradients when the input values are very small.

- Moved epsilon inside the square root calculation instead of after it
- Increased default from 1e-10 to 1e-8 for better stability
- Added test

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.

### CODE REVIEW

File: monai/losses/perceptual.py [hunk 1]
```diff
@@ -271,8 +271,8 @@ def spatial_average_3d(x: torch.Tensor, keepdim: bool = True) -> torch.Tensor:
     return x.mean([2, 3, 4], keepdim=keepdim)
 
 
-def normalize_tensor(x: torch.Tensor, eps: float = 1e-10) -> torch.Tensor:
-    norm_factor = torch.sqrt(torch.sum(x**2, dim=1, keepdim=True))
+def normalize_tensor(x: torch.Tensor, eps: float = 1e-8) -> torch.Tensor:
+    norm_factor = torch.sqrt(torch.sum(x**2, dim=1, keepdim=True) + eps)
     return x / (norm_factor + eps)
```
Comment (ericspod) [inline monai/losses/perceptual.py:276]:
```suggestion
def normalize_tensor(x: torch.Tensor, eps: float = 1e-8) -> torch.Tensor:
    norm_factor = torch.sqrt(torch.sum(x**2, dim=1, keepdim=True) + eps)
    return x / norm_factor
```

Do we want to remove `eps` from the denominator? As proposed `eps` will contribute twice to the final result.

File: tests/test_perceptual_loss_stability.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    def tearDown(self):
+        set_determinism(None)
+
+    @parameterized.expand([["e-3", 1e-3], ["e-6", 1e-6], ["e-9", 1e-9], ["e-12", 1e-12]])  # Small values
+    def test_normalize_tensor_stability(self, name, scale):
+        """Test that small values don't produce NaNs + are handled gracefully."""
+        # Create tensor
+        x = torch.zeros(2, 3, 10, 10, requires_grad=True)
+
+        optimizer = optim.Adam([x], lr=0.01)
```
Comment (ericspod) [inline tests/test_perceptual_loss_stability.py:38]: I don't think the optimizer is needed for this test?

File: tests/test_perceptual_loss_stability.py
[the remark this answers is not in the document]
Comment (cvbourne) [reply tests/test_perceptual_loss_stability.py:1]: Roger.

File: tests/test_perceptual_loss_stability.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        x_scaled = x * scale
```
[the remark this answers is not in the document]
Comment (cvbourne) [reply tests/test_perceptual_loss_stability.py:39]: Will add a comment.

File: tests/test_perceptual_loss_stability.py (hunk 2 above)
Comment (cvbourne) [reply tests/test_perceptual_loss_stability.py:38]: Not needed, will remove.

File: tests/test_perceptual_loss_stability.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (virginiafdez) [reply tests/test_perceptual_loss_stability.py:39]: I don't understand the point of this test with regards to the next one; instead of a zeros tensor, couldn't it be a random one which will be then multiplied by a really small number?

**Review by virginiafdez — COMMENTED** [review]
The changes look good to me. I'd modify one of the tests, but the rest is fine.

**Review by virginiafdez — DISMISSED** [review]
Besides my comment about the point of one of the tests, I think this PR can be merged, as long as the errors happening on the automatic tests are fixed.

**Comment by ericspod** [comment]
Hi @cvbourne if you can come back to this one and address the issues raised by us and Coderabbit, we'd still like to merge this one. Thanks!

**Comment by ericspod** [comment]
I think this PR has been superseded now so closing.
